Added autotable search

MillanWangGadget opened the pull request:
- **UPDATE**
  - Added searching functionality to the auto table
    - This is done by using the existing fuzzy search functionality in the useFindMany hook and incorporating this into the existing query system in the Polaris table
  - Updated the response type of the `useTable` hook
    - Keeping it as a `position dependant` array is annoying if you want a late-in-the-array value and don't want the middle-of-the-array values. You'd need to count the ignored properties

## PR Checklist

- [x] Important or complicated code is tested
- [ ] Any user facing changes are documented in the Gadget-side changelog
- [ ] Any immediate changes are slated for release in Gadget via a generated package dependency bump
- [ ] Versions within this monorepo are matching and there's a valid upgrade path

danroberts commented:
`pdated the response type of the useTable hook
Keeping it as a position dependant array is annoying if you want a late-in-the-array value and don't want the middle-of-the-array values. You'd need to count the ignored properties`

I'm not sure I follow 100%, but I think its good to keep the convention of the other hooks, which is usually [{variables}, refresh]. The {variables} which are destructured are not position dependent.

danroberts reviewed (commented):
As part of this, could you also see if you can find a nice way in the UI to 'exit' the search bar? Maybe on blur?

MillanWangGadget commented on packages/react/spec/auto/PolarisAutoTable.stories.jsx:
```diff
@@ -14,7 +15,7 @@ export default {
       return (
         <Provider api={api}>
-          <AppProvider>
+          <AppProvider i18n={translations}>
```
This is required to get the text into the cancel button. Unfortunately, Shopify does not have a default i18n language which leads to empty strings when one is not explicitly given

MillanWangGadget commented:
> I'm not sure I follow 100%, but I think its good to keep the convention of the other hooks, which is usually [{variables}, refresh]. The {variables} which are destructured are not position dependent.

Updated such that the first array parameter has `{table:{rows, columns, etc}, search:{value, set,clear,}}` 


> As part of this, could you also see if you can find a nice way in the UI to 'exit' the search bar? Maybe on blur?

Updated to add a cancel button which also clears the search

danroberts commented on packages/react/src/useTable.tsx:
```diff
@@ -45,20 +45,35 @@ export interface TableOptions {
-  (opts?: Partial<OperationContext>) => void
+  {
+    table: TableData<Data> & {
+      page: TablePagination;
```
I don't think I'm 100% on board for grouping these under a table key -- I think it just makes it a bit more annoying to destructure them when you're consuming the hook.

I do think all the search props can be grouped together tho.

MillanWangGadget replied:
Updated such that the search stuff is grouped together and the previous variables are untouched

danroberts reviewed (commented):
I also think we need to debounce the input a bit -- making a network request on each keypress seems aggressive -- do you see any opportunity to debounce by ~500ms ish?

MillanWangGadget commented:
> I also think we need to debounce the input a bit -- making a network request on each keypress seems aggressive -- do you see any opportunity to debounce by ~500ms ish?

Updated to use debounce system for the search. I set the timeout to 300ms cause I felt like (500ms+ the request time) felt a little bit slow. This is easily adjusted in the call to `debounce` though 👍

MillanWangGadget commented:
Merging now because test failure is in an unrelated form matter
